src/network/networkd-address.c

```diff
@@ -658,9 +664,100 @@ int address_configure(
                 return log_link_error_errno(link, r, "Could not add address: %m");
         }
 
+        if (address->ipv4_duplicate_address_detection) {
+                if (DEBUG_LOGGING) {
+                        _cleanup_free_ char *pretty = NULL;
+
+                        (void) in_addr_to_string(AF_INET, &address->in_addr, &pretty);
+                        log_debug("Starting IPv4ACD client. Probing address '%s'", strna(pretty));
+                }
+
+                r = sd_ipv4acd_start(address->acd);
+                if (r < 0)
+                        log_link_warning_errno(link, r, "Failed to start IPv4ACD client, ignoring: %m");
+        }
+
         return 1;
 }
 
+static void static_address_on_acd(sd_ipv4acd *acd, int event, void *userdata) {
+        _cleanup_free_ char *pretty = NULL;
+        Address *address;
+        Link *link;
+        int r;
+
+        assert(acd);
+        assert(userdata);
+
+        address = (Address *) userdata;
+        link = address->link;
+
+        (void) in_addr_to_string(AF_INET, &address->in_addr, &pretty);
+        switch (event) {
+
+        case SD_IPV4ACD_EVENT_STOP:
+                log_link_debug(link, "Stopping ACD client ...");
+                return;
+
+        case SD_IPV4ACD_EVENT_BIND:
+                log_link_debug(link, "Successfully claimed address '%s'", strna(pretty));
+                link_check_ready(link);
+                break;
+
+        case SD_IPV4ACD_EVENT_CONFLICT:
+                log_link_warning(link, "DAD conflict. Dropping address '%s'", strna(pretty));
+                r = address_remove(address, link, NULL);
+                if (r < 0)
+                        log_link_error_errno(link, r, "Failed to drop DAD conflicted address '%s'", strna(pretty));;
+
+                link_check_ready(link);
+                break;
+
+        default:
+                assert_not_reached("Invalid IPv4ACD event.");
+        }
+
+        sd_ipv4acd_stop(acd);
+
+        return;
+}
+
+int configure_ipv4_duplicate_address_detection(Link *link, Address *address) {
+        int r;
+
+        assert(link);
+        assert(address);
+        assert(!address->link && address->network);
+
+        address->link = link_ref(link);
+
+        r = sd_ipv4acd_new(&address->acd);
+        if (r < 0)
+                return log_oom();
+
+        r = sd_ipv4acd_attach_event(address->acd, NULL, 0);
+        if (r < 0)
+                return log_link_error_errno(link, r, "IPv4ACD: Failed to attch event: %m");
```

Comment:
Typo here. But maybe remove all those specific error messages. AFter all, we don't expect them to ever fail. So this function could just do `return r`, and the caller could do:
```C
                       r = configure_ipv4_duplicate_address_detection(link, address);
                        if (r < 0)
                                return log_link_error_errno(link, r, "IPv4ACD: Failed to configure DAD: %m");
```